File: algorithms/leetcode/454-4sum-ii/4sum-ii.cpp
#include <unordered_map>
#include <vector>
using namespace std;

class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        int N = A.size();

        int result = 0;

        unordered_map<int, int> m;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                m[A[i] + B[j]]++;
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                int sum = -C[i] -D[j];

                if (m.find(sum) != m.end()) {
                    result += m[sum];
                }
            }
        }

        return result;
    }
};

File: leetcode/454-4sum-ii/4sum-ii.js
/**
 * @param {number[]} A
 * @param {number[]} B
 * @param {number[]} C
 * @param {number[]} D
 * @return {number}
 */
const fourSumCount = function (A, B, C, D) {
  const map = new Map();

  for (const a of A) {
    for (const b of B) {
      const sum = a + b;
      if (map.has(sum)) {
        map.set(sum, map.get(sum) + 1);
      } else {
        map.set(sum, 1);
      }
    }
  }

  let result = 0;
  for (const c of C) {
    for (const d of D) {
      const sum = c + d;
      if (map.has(-sum)) {
        result += map.get(-sum);
      }
    }
  }

  return result;
};
